Project: Fake Quantization

ranbir-sharma opened the pull request:
***Project : Fake Quantization***

Project Details are as follow

- Fake Quantization is added to the LLTFI to ensure fault injection within the targeted layers specified in the yaml files 
- Lib code is added to facilitate the process for maintaining and Quantizing / Dequantizing the values as provided 
- CMake Files are updated to build the code for Fake Quantization while setting up LLTFI

ranbir-sharma committed:
Added the code for FakeQuantization project

ranbir-sharma committed:
Added the lib code to build the files needed for FakeQuantization

abrahamchan commented:
Do you have any tests for this feature? For example, it would be great to have an example under [sample_programs/ml_sample_programs/](https://github.com/DependableSystemsLab/LLTFI/tree/master/sample_programs/ml_sample_programs/) to demonstrate the use of fake quantization.

ranbir-sharma committed:
Added documentation and error checking for fake quantization

ranbir-sharma commented:
> Do you have any tests for this feature? For instance, it would be great to have an example under [sample_programs/ml_sample_programs/](https://github.com/DependableSystemsLab/LLTFI/tree/master/sample_programs/ml_sample_programs/) to demonstrate the use of fake quantization.
> 
> Also, the [input_masterlist_ml.yaml](https://github.com/DependableSystemsLab/LLTFI/blob/master/docs/input_masterlist_ml.yaml) should be updated with the `fakeQuant` option.

Added Documentation in the `input_masterlist_ml.yaml` with examples and added error checks for fake Quantization in `instrument.py`

abrahamchan commented on llvm_passes/core/FakeQuantizationPass.cpp:
```diff
@@ -0,0 +1,515 @@
+#include <unordered_set>
+
+#include "llvm/Support/CommandLine.h"
+// #include "Controller.h"
```
Please delete this include if not required.

ranbir-sharma replied:
Resolved!

ranbir-sharma committed:
Added revisions in accordance to the PR comments

ranbir-sharma commented on llvm_passes/core/FakeQuantizationPass.h:
```diff
@@ -0,0 +1,43 @@
+
+} // namespace llfi
+
+#endif // LLFI_H
```
Resolved!

ranbir-sharma commented on bin/instrument.py:
```diff
@@ -317,6 +319,26 @@ def readCompileOption():
+    assert 0 < bitWidth, "BitWidth must be a integer greater than 0"
+    assert math.log2(bitWidth).is_integer(), "BitWidth must be a exponent of power 2"
+    assert 0 <= minPercentileOutlierThreshold and minPercentileOutlierThreshold <= 100, "Minimum Percentile Value for Percentile should be greater than or equal to 0"
+    assert 0 <= maxPercentileOutlierThreshold and maxPercentileOutlierThreshold <= 100, "Maximum Percentile Value for Percentile should be greater than or equal to 0 and lesser than or equal to 100"
```
Resolved!

ranbir-sharma commented on runtime_lib/FakeQuantizationLib.h:
```diff
@@ -0,0 +1,24 @@
+}
+#endif
+
+#endif // FAKE_QUANTIZATION_H
```
Resolved!

ranbir-sharma commented on llvm_passes/core/FakeQuantizationPass.h:
```diff
@@ -0,0 +1,43 @@
+
+} // namespace llfi
+
+#endif // LLFI_H
```
Resolved!

ranbir-sharma commented on runtime_lib/FakeQuantizationLib.h:
```diff
@@ -0,0 +1,24 @@
+extern "C" {
+#endif
+float getWAndX(float w1, float x1, int currentLayerIndex, int totalNumberOfLayers);
+void finished(int currentLayerIndex, int totalNumberOfLayers, int minPercentileThreshold, int maxPercetileThreshold, int bitWidth);
```
Resolved!

ranbir-sharma commented on runtime_lib/FakeQuantizationLib.h:
```diff
@@ -0,0 +1,24 @@
+float Quantize(float w1, float x1, int currentLayerIndex, int totalNumberOfLayers);
+float QuantizeMatMul(float w1, float x1, int currentLayerIndex, int totalNumberOfLayers);
+float FakeQuantIntegerBasedAddition(float num1, float num2);
+float FakeQunatDequnatizeAndBiasAddition(float num1, float num2);
```
Resolved!

ranbir-sharma commented on docs/input_masterlist_ml.yaml:
```diff
@@ -139,3 +139,25 @@ runOption:
+# Fake Qunatization parameters comes within compileOption. In here, you need to mention the targer layer - conv or matmul (currently only supporting these)
+# minPercentileOutlierThreshold and maxPercentileOutlierThreshold are optional parameters. On not mentioning these parameters, they are autoamtically set to 
+# minPercentileOutlierThreshold = 0 and maxPercentileOutlierThreshold = 100
+# bitWidth needs to be mentioned and is the bit width of the Quantized Int numbers which needs to be a exponent of power 2
```
Resolved!

ranbir-sharma commented on docs/input_masterlist_ml.yaml:
```diff
@@ -139,3 +139,25 @@ runOption:
+      bitWidth: 8
+      
+
+# Fake Qunatization parameters comes within compileOption. In here, you need to mention the targer layer - conv or matmul (currently only supporting these)
```
Resolved!

abrahamchan commented on runtime_lib/FakeQuantizationLib.cpp:
```diff
@@ -0,0 +1,427 @@
+        }
+        // printf("This is going away - %i with w - %f with Q(w) - %i and x - %f with Q(x) - %i\n", ans, w1, quantize_helper(w1, s_w), x1, quantize_helper(x1, s_x));
+        return ans;
+        // return 8.0;
```
Remove this code.

abrahamchan commented on runtime_lib/FakeQuantizationLib.cpp:
```diff
@@ -0,0 +1,427 @@
+        }
+        // printf("This is going away - %i with w - %f with Q(w) - %i and x - %f with Q(x) - %i\n", ans, w1, quantize_helper(w1, s_w), x1, quantize_helper(x1, s_x));
+        return ans;
+        // return 8.0;
```
Remove this code.

abrahamchan commented on runtime_lib/FakeQuantizationLib.cpp:
```diff
@@ -0,0 +1,427 @@
+        max_w = find_percentile_value(w_values, w_index, maxPercentileThreshold);
+    }
+
+    // min_w = find_percentile_value(w_values, w_index, lowerPercentile);
```
Remove this section of commented out code.

abrahamchan commented on runtime_lib/FakeQuantizationLib.cpp:
```diff
@@ -0,0 +1,427 @@
+{
+    int intNum1 = (int)num1;
+    int intNum2 = (int)num2;
+    // printf("Got this in Addition - %i and %i and their sum %i\n", intNum1, intNum2, intNum1 + intNum2);
```
Remove this debugging log.

abrahamchan commented on runtime_lib/FakeQuantizationLib.cpp:
```diff
@@ -0,0 +1,427 @@
+        {
+            ans = (-1 * max_number);
+        }
+        // printf("This is going away - %i with w - %f with Q(w) - %i and x - %f with Q(x) - %i\n", ans, w1, quantize_helper(w1, s_w), x1, quantize_helper(x1, s_x));
```
Remove this debugging log.

abrahamchan commented on runtime_lib/FakeQuantizationLib.cpp:
```diff
@@ -0,0 +1,427 @@
+    int QunatizeIntNum = (int)QunatizeNum;
+    float dequnatizedFloat = dequantize(QunatizeIntNum);
+    float result = dequnatizedFloat + Basis;
+    // printf("Returning this - %f\n", result);
```
Remove these prints used for debugging.

abrahamchan reviewed (changes requested):
Could you please make a second round of changes? I didn't catch the other typos in the previous review.

ManiSadati commented:
Does this PR still have the unexpected large value issue for all the models?

ranbir-sharma committed:
Added revisions

ranbir-sharma committed:
Added revisions - II

abrahamchan commented:
LGTM!
